### PR TITLE
Fix non-AVX fallback paths (c128 SIGILL + f64 sigmoid accuracy)

### PR DESCRIPTION
**Pull request by tphakala**
## Summary

Two fallback-correctness bugs that only manifest on x86 CPUs without AVX, found
during a fallback audit and confirmed empirically under qemu. Every CI and dev
host has AVX, so these paths were never exercised and the bugs reached `main`.

**1. `c128` SSE2 kernels SIGILL on CPUs without SSE4.1.** The c128 init switch
dispatches the `*SSE2` kernels on `cpu.X86.SSE2` (the amd64 baseline), but five
of them used post-SSE2 instructions:

- `mulSSE2` / `mulConjSSE2` / `scaleSSE2` used `BLENDPD` (SSE4.1)
- `absSSE2` / `absSqSSE2` used `MOVDDUP` (SSE3)

On a CPU advertising no SSE4.1 (e.g. Core 2 / Conroe, or a default `qemu64` VM)
the `BLENDPD` kernels fault with SIGILL; on a genuine SSE2-only CPU the
`MOVDDUP` kernels fault too. Replaced with bit-identical SSE2 forms:
`BLENDPD $0x02, Xs, Xd` -> `SHUFPD $0x2, Xs, Xd` (both yield `[Xd.lo, Xs.hi]`),
and `MOVDDUP X0, X1` -> `MOVAPD X0, X1` (only the low lane is stored). `c64` was
already correct (it gates its similarly named kernels on `cpu.X86.SSE41`).

**2. `f64` sigmoid scalar fallback diverged from the SIMD kernel.** `sigmoid64Go`
hard-clamped to `0` for `x < -20` (a leftover from the pre-#33 rational
approximation), while the AVX/AVX-512 kernels compute the accurate
`1/(1+e^-x)`. `TestSigmoidAccuracy` asserts that accuracy across `-30..30` but
only runs the dispatched path, so on AVX2 CI the inaccurate fallback was never
checked. `sigmoid(-25)` is ~1.4e-11, not 0. Now computes `1/(1+e^-x)` with the
exponent clamped to `+/-expOverflowThreshold` (no overflow to +Inf), matching
the kernels within tolerance.

## Regression guards

- `TestSSE2KernelsAreBaselineSSE2` (c128): scans the `*SSE2` TEXT blocks and
  fails on any post-SSE2 mnemonic. Runs on normal CI, no qemu needed.
- `TestSigmoid64Go_NoFlushToZero` (f64): calls the scalar fallback directly so
  it runs on every host, unlike the dispatched accuracy test.
- New `cross-isa` CI job: runs the amd64 suite under `qemu-x86_64 -cpu Conroe`
  (SSSE3, no SSE4.1, no AVX) and a stripped SSE2-only `qemu64` model. qemu v7+
  enforces the `-cpu` feature set and raises #UD for out-of-baseline ops, so
  these models reproduce real-hardware faults. This is what would have caught
  both bugs.

## Test Plan

- [x] `go test -race ./...` passes natively
- [x] `go vet ./...` clean for amd64 and arm64
- [x] `golangci-lint run ./...` clean
- [x] Full suite passes under `qemu-x86_64` `-cpu Conroe`, SSE2-only, `Nehalem`, `Penryn` (all no-AVX)
- [x] c128 numerical results bit-identical to before (verified native)
- [x] Regression tests verified red against the old code, green after the fix

## Summary by CodeRabbit

* **New Features**
  * Added QEMU-based cross-ISA CPU compatibility testing to validate system behavior across processors without AVX and extended SSE instruction support.

* **Improvements**
  * Enhanced sigmoid function numerical stability through exponent-clamping logic, replacing hard cutoffs with smooth saturation behavior for extreme input values.
  * Optimized assembly implementations for improved instruction compatibility and performance.

* **Tests**
  * Added baseline validation test and regression test to prevent numerical edge-case issues.

### CODE REVIEW

File: f64/f64_go.go [hunk 1]
```diff
@@ -452,16 +451,21 @@ func sigmoid64Go(dst, src []float64) {
...
-		case x < -sigmoidClampThreshold:
-			dst[i] = 0.0
-		default:
-			dst[i] = 1.0 / (1.0 + math.Exp(-x))
+		case z > expOverflowThreshold:
+			z = expOverflowThreshold
+		case z < -expOverflowThreshold:
+			z = -expOverflowThreshold
 		}
+		dst[i] = 1.0 / (1.0 + math.Exp(z))
```
[the remark this answers is not in the document]
Comment (tphakala) [reply f64/f64_go.go:468]: Keeping the clamp. I probed the dispatched AVX kernel: for x <= -709 it returns 1.2167807515655636e-308 (= 1/(1+exp(709))), not 0 -- it clamps the exponent to 709. The unclamped 1/(1+exp(-x)) returns 0 for x <= -710, which re-introduces a SIMD-vs-scalar divergence at the tail, the exact bug class this PR fixes. The clamped fallback stays consistent with the kernel (both ~1.2e-308). math.Exp overflowing to +Inf is safe (no NaN), but the resulting 0 disagrees with the kernel, so I am not removing the clamp.

File: f64/go_impl_newops_test.go [hunk 2]
```diff
@@ -71,23 +71,49 @@ func TestSubFromScalarGo_EmptyDst(_ *testing.T) {
...
+	// With z = -x clamped to expOverflowThreshold, sigmoid(-large) is
+	// 1/(1+exp(709)) ~ 1.2e-308: vanishingly small but strictly positive.
+	wantNeg := 1.0 / (1.0 + math.Exp(expOverflowThreshold))
+	if dst[1] != wantNeg {
+		t.Errorf("sigmoid64Go(-large) = %v, want %v", dst[1], wantNeg)
 	}
 	if dst[2] != 0.5 {
 		t.Errorf("sigmoid64Go(0) = %v, want 0.5", dst[2])
 	}
 }
```
[the remark this answers is not in the document]
Comment (tphakala) [reply f64/go_impl_newops_test.go:93]: Following from the above: with the clamp retained, sigmoid64Go(-(709+1)) is 1/(1+exp(709)) ~ 1.2e-308, not 0, which matches the AVX kernel. The test asserts that value (wantNeg) on purpose; changing it to expect 0.0 would lock in the divergent behavior. Leaving as is.

File: c128/sse2_baseline_test.go [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+		}
+
+		mnemonic := strings.Fields(line)[0]
+		if strings.HasSuffix(mnemonic, ":") { // label
+			continue
+		}
+		if isa, bad := postSSE2Mnemonics[mnemonic]; bad {
+			t.Errorf("%s:%d: %s uses %s (%s); c128 dispatches *SSE2 kernels on plain SSE2, so this SIGILLs on CPUs without %s",
+				asmFile, lineNo+1, curFunc, mnemonic, isa, isa)
+		}
```
[the remark this answers is not in the document]
Comment (tphakala) [reply c128/sse2_baseline_test.go]: Done in ad6d3a0. Added V (VEX/EVEX) and K (AVX-512 opmask) prefix checks ahead of the named denylist; no SSE2 mnemonic or Plan9 pseudo-op starts with V or K, so it has no false positives on the current kernels.

File: .github/workflows/ci.yml [hunk 4]
```diff
@@ -127,3 +127,50 @@ jobs:
...
+    name: Cross-ISA (no-AVX fallbacks)
+    runs-on: ubuntu-latest
+    steps:
+      - name: Checkout code
+        uses: actions/checkout@v6
+
+      - name: Set up Go
+        uses: actions/setup-go@v6
+        with:
+          go-version: ${{ env.GO_VERSION }}
```
[the remark this answers is not in the document]
Comment (tphakala) [reply .github/workflows/ci.yml:147]: Fixed the error-swallowing in ad6d3a0: the job now detects no-test packages via go list -f '{{len .TestGoFiles}} {{len .XTestGoFiles}}' and runs go test -c without || true or the stderr redirect, so a genuine compile failure fails the job. I did not pin the actions to commit SHAs or add persist-credentials: false: every other job here uses @v6 tags and the workflow token is already globally scoped to contents: read, so adding those to only this job would be inconsistent. Supply-chain pinning would be better as a separate workflow-wide change.

File: .github/workflows/ci.yml [hunk 5]
```diff
@@ -127,3 +127,50 @@ jobs:
...
+          set -uo pipefail
+          # Conroe: SSSE3, no SSE4.1, no AVX (catches the c128 BLENDPD SIGILL).
+          # qemu64 stripped to SSE2: the amd64 baseline (catches SSE3+ ops like
+          # MOVDDUP and any non-SSE2 instruction in an SSE2-gated kernel).
+          models=("Conroe" "qemu64,-pni,-ssse3,-sse4.1,-sse4.2,-avx,-avx2")
+          fail=0
+          for pkg in $(go list ./...); do
+            bin="/tmp/$(echo "$pkg" | tr '/.' '__').test"
+            CGO_ENABLED=0 go test -c -o "$bin" "$pkg" 2>/dev/null || true
+            [ -f "$bin" ] || continue   # package has no test files
```
[the remark this answers is not in the document]
Comment (tphakala) [reply .github/workflows/ci.yml]: Fixed in ad6d3a0: removed || true and the stderr redirect so compile failures fail the job, and packages without tests are skipped up front via go list.

**Comment by tphakala** [comment]
@coderabbitai review

File: f64/f64_go.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (tphakala) [reply f64/f64_go.go:468]: Same conclusion as the thread above: the clamp stays. The dispatched AVX/AVX-512 kernel clamps the exponent to 709 and returns ~1.2e-308 for x <= -709, not 0 (probed on this host). Removing the clamp makes the scalar fallback return 0 there, which re-introduces the scalar-vs-SIMD divergence this PR exists to fix. sigmoid being bounded in [0,1] is not the concern; matching the kernel within tolerance is. Keeping it.

File: f64/go_impl_newops_test.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (tphakala) [reply f64/go_impl_newops_test.go:93]: The test intentionally uses expOverflowThreshold + 1 to exercise the z > threshold clamp branch (and -(threshold+1) for the z < -threshold branch), because the clamp is being kept. Using threshold - 1 would skip the clamp branches entirely and defeat the test's purpose. Smooth mid-range behaviour is already covered by TestSigmoid64Go_NoFlushToZero.
